test/puppetlabs/puppetdb/generative/generators.clj

```diff
@@ -0,0 +1,297 @@
+(ns puppetlabs.puppetdb.generative.generators
+  (:require [clojure.test.check.generators :as gen]
+            [com.gfredericks.test.chuck.generators :as chuck-gen]
+            [clj-time.core :as time :refer [now]]))
+
+
+;;; common data types and combinators
+(defn nilable [g]
+  (gen/one-of [(gen/return nil) g]))
+
+(defn string+
+  ([]
+   (gen/fmap clojure.string/join (gen/vector gen/char-alphanumeric 1 255)))
+  ([max-len]
+   (gen/fmap clojure.string/join (gen/vector gen/char-alphanumeric 1 max-len))))
+
+(defn pg-smallint [] (gen/choose -32768 32767))
+(defn pg-integer [] (gen/choose -2147483648 2147483647))
+(defn pg-pos-integer [] (gen/choose 1 2147483647))
+(defn pg-bigint [] (gen/choose -9223372036854775808 9223372036854775807))
+
+(defn uuid-string [] (gen/fmap str gen/uuid))
+
+(defn datetime []
+  (gen/fmap clj-time.coerce/from-long
+            (gen/choose 50000000000 2000000000000)))
+
+
+
+;;; catalog generator
+
+(defn certname [] (string+))
+(defn environment [] (string+))
```

Comment:
I think the main generator definition here should be a the definition of what the environment /can/, be, i.e. keep it close to what a spec would be. But I can see doing that in the overrides. Not sure if test.check knows how to do that kind of thing, but I'm sure we could teach it.